### PR TITLE
chore: Cleanup

### PR DESCRIPTION
**Pull request by arv**

### CODE REVIEW

**Comment by arv** [comment]
Consider this an extensive code review.

I want the sample code to be our best foot forward and I wanted to set some precedence here.

This repo is missing eslint which should be setup and use the same config as replicache-internal. Maybe we should create a `@rocicorp/eslint` repo? A github repo should be sufficient. No need to publish to npm (I believe).

Comments and motivations in the self review

File: ts-web-component/src/component/my-todo.ts [hunk 1]
```diff
@@ -1,149 +1,155 @@
...
+  private _listContainer: Element | null = null;
+  private _newTodoInput: HTMLInputElement | null = null;
+  private _itemCount: Element | null = null;
+  private _filterLinks: Iterable<Element> = [];
+  private _list: Todo[] = [];
+  private _filteredList: Todo[] = [];
+  private _filter: Filter = "all";
+  private _r: Replicache<M> | null = null;
+
+  private async _createTodo(text: string) {
```
[the remark this answers is not in the document]
Comment (arv) [reply ts-web-component/src/component/my-todo.ts:48]: I renamed this to `createX` and kept all the `handleX` as event callbacks.

File: ts-web-component/src/component/my-todo.ts [hunk 2]
```diff
@@ -1,149 +1,155 @@
...
       },
     });
-    this.appendChild(templateTodo.content.cloneNode(true));
-    this.newTodoInput = this.querySelector(
-      "#new-todo"
-    ) as HTMLInputElement | null;
-    this.listContainer = this.querySelector("#list-container");
-    this.itemCount = this.querySelector("#itemCount");
-    this.filter = this.querySelectorAll(".filters a");
-    this.newTodoInput!.addEventListener("keyup", (e: KeyboardEvent) => {
```
Comment (arv) [inline ts-web-component/src/component/my-todo.ts:109]: All the other event handlers where "methods". Changed this one to a method as well... but inline arrow functions are fine if we never need to remove the event listener.

File: ts-web-component/src/component/my-todo.ts [hunk 3]
```diff
@@ -1,149 +1,155 @@
...
+    this._newTodoInput = this.querySelector(".new-todo");
+    this._listContainer = this.querySelector(".todo-list");
+    this._itemCount = this.querySelector(".item-count");
+    this._filterLinks = this.querySelectorAll(".filters a");
+    assert(this._newTodoInput);
+    this._newTodoInput.addEventListener("keyup", this._handleTodoInputKeyUp);
+
+    for (const filter of this._filterLinks) {
+      filter.addEventListener("click", this._handleFilterClick);
+    }
```
Comment (arv) [inline ts-web-component/src/component/my-todo.ts:107]: I changed the order here to ensure the DOM is here when subscribe is called.

File: ts-web-component/src/component/my-todo.ts [hunk 4]
```diff
@@ -1,149 +1,155 @@
...
-    });
+  }
 
-    if (this.filter.length) {
-      this.filter.forEach((filter) => {
-        filter.addEventListener("click", this.handleFilterClick.bind(this));
-      });
-    }
+  disconnectedCallback() {
+    this._r?.close();
```
Comment (arv) [inline ts-web-component/src/component/my-todo.ts:121]: important or we keep things around forever

File: ts-web-component/src/component/my-todo.ts [hunk 5]
```diff
@@ -1,149 +1,155 @@
...
-      $item.setAttribute("checked", item.completed.toString());
-      $item.todoID = item.id;
-      $item.addEventListener("onRemove", this.handleDeleteTodos.bind(this));
-      $item.addEventListener("onToggle", this.handleUpdateTodo.bind(this));
-      this.listContainer?.appendChild($item);
-    });
+  private _render() {
+    assert(this._listContainer);
+    assert(this._itemCount);
+    this._itemCount.textContent = `${this._list.length}`;
```
Comment (arv) [inline ts-web-component/src/component/my-todo.ts:127]: use textContent for text

File: ts-web-component/src/component/my-todo.ts [hunk 6]
```diff
@@ -1,149 +1,155 @@
...
   }
 
-  disconnectedCallback() {}
-
-  _render() {
-    if (!this.listContainer) return;
-    this.itemCount!.innerHTML = `${this._list.length}`;
-    this.listContainer.innerHTML = "";
-    this._filteredList.forEach((item: Todo) => {
-      let $item: TodoItemElement = document.createElement(
```
Comment (arv) [inline ts-web-component/src/component/my-todo.ts:130]: The name if `$item` is not consistent. eslint should catch this.

File: ts-web-component/src/component/my-todo.ts [hunk 7]
```diff
@@ -1,149 +1,155 @@
...
-        "todo-item"
-      ) as TodoItemElement;
-      $item.setAttribute("text", item.text);
-      $item.setAttribute("checked", item.completed.toString());
-      $item.todoID = item.id;
-      $item.addEventListener("onRemove", this.handleDeleteTodos.bind(this));
```
Comment (arv) [inline ts-web-component/src/component/my-todo.ts:136]: binding arrow functions does nothing

File: ts-web-component/src/component/my-todo.ts [hunk 8]
```diff
@@ -1,149 +1,155 @@
...
-      if (filter.id === this._filter) {
-        filter.classList.add("selected");
-      } else {
-        filter.classList.remove("selected");
-      }
-    });
+      this._listContainer.appendChild(item);
+    }
+
+    for (const filter of this._filterLinks) {
```
Comment (arv) [inline ts-web-component/src/component/my-todo.ts:140]: prefer for of loops over forEach

File: ts-web-component/src/component/todo-item.ts [hunk 9]
```diff
@@ -11,35 +12,37 @@ templateTodoItem.innerHTML = `
...
+export interface TodoItemEventHandlers {
+  // This is not really correct but good enough for now.
+  addEventListener<T>(
+    type: string,
+    listener: (e: CustomEvent<T>) => void
+  ): void;
+}
+
+export class TodoItem extends HTMLElement {
+  static observedAttributes = ["text", "completed"];
```
Comment (arv) [inline ts-web-component/src/component/todo-item.ts:24]: no need for a getter. Maybe cargo culted from somewhere?

File: ts-web-component/src/component/todo-item.ts [hunk 10]
```diff
@@ -11,35 +12,37 @@ templateTodoItem.innerHTML = `
...
   connectedCallback() {
     this.appendChild(templateTodoItem.content.cloneNode(true));
-    this.item = this.querySelector(".item");
-    this.removeButton = this.querySelector(".destroy");
-    this.text = this.querySelector("label");
-    this.checkbox = this.querySelector("input");
-    this.removeButton?.addEventListener("click", (e: Event) => {
-      e.preventDefault();
+    this._item = this.querySelector(".item");
+    const removeButton = this.querySelector(".destroy");
```
Comment (arv) [inline ts-web-component/src/component/todo-item.ts:34]: remove button was only used in this function

File: ts-web-component/src/component/todo-item.ts [hunk 11]
```diff
@@ -54,31 +57,44 @@ export default class TodoItem extends HTMLElement {
     return this._todoID;
   }
 
-  disconnectedCallback() {}
+  get text() {
+    return this.getAttribute("text") ?? "";
```
Comment (arv) [inline ts-web-component/src/component/todo-item.ts:61]: One of the nice things about custom elements/web components is that you can let the DOM hold the source of truth.

File: ts-web-component/src/component/todo-item.ts [hunk 12]
```diff
@@ -54,31 +57,44 @@ export default class TodoItem extends HTMLElement {
...
-  attributeChangedCallback(_name: any, _oldValue: any, newValue: string) {
-    if (_name === "text") {
-      this._text = newValue;
-    } else if (_name === "checked") {
-      this._completed = newValue === "true";
+  set completed(v: boolean) {
+    if (v) {
+      this.setAttribute("completed", "");
+    } else {
+      this.removeAttribute("completed");
```
Comment (arv) [inline ts-web-component/src/component/todo-item.ts:76]: This is called a boolean attribute in DOM speak. The presence means true and the absence means false

File: ts-web-component/src/index.ts [hunk 13]
```diff
@@ -1,5 +1,5 @@
-import MyTodo from "./component/my-todo";
-import TodoItem from "./component/todo-item";
+import { MyTodo } from "./component/my-todo";
```
Comment (arv) [inline ts-web-component/src/index.ts:1]: Default exports are considered bad...

File: ts-web-component/src/setup-replicache.ts [hunk 14]
```diff
@@ -24,10 +25,12 @@ export async function setupReplicache(): Promise<Replicache<M>> {
     window.location.href = "/listId/" + listId;
   }
 
+  // See https://doc.replicache.dev/licensing for how to get a license key.
+  const licenseKey = import.meta.env.VITE_REPLICACHE_LICENSE_KEY;
+  assert(licenseKey, "Missing VITE_REPLICACHE_LICENSE_KEY");
+
   const r = new Replicache<M>({
-    // See https://doc.replicache.dev/licensing for how to get a license key.
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (arv) [inline ts-web-component/src/setup-replicache.ts:29]: eslint was not configured so this didn't do anything.

Changed it to a runtime assert instead

File: ts-web-component/src/space.ts [hunk 15]
```diff
@@ -1,27 +1,22 @@
...
-  return await res.json();
+  return await fetchJSON("spaceExists", spaceID);
 }
 
 export async function createSpace(spaceID?: string): Promise<any> {
-  const res = await fetch("/api/replicache/createSpace", {
+  return await fetchJSON("createSpace", spaceID);
+}
+
+async function fetchJSON(apiName: string, spaceID: string | undefined) {
```
Comment (arv) [inline ts-web-component/src/space.ts:9]: share code

File: ts-web-component/src/style.css [hunk 16]
```diff
@@ -302,6 +302,7 @@ body {
 }
 
 .filters li a {
+  cursor: pointer;
```
Comment (arv) [inline ts-web-component/src/style.css:305]: less repetition this way

File: ts-web-component/src/types.ts [hunk 17]
```diff
@@ -1,4 +0,0 @@
-export type TodoItemElement = HTMLInputElement & {
```
Comment (arv) [inline ts-web-component/src/types.ts:1]: It was not an input element

but keep the types together with their implementations

File: ts-web-component/tsconfig.json [hunk 18]
```diff
@@ -6,7 +6,7 @@
     "allowJs": false,
     "skipLibCheck": true,
     "esModuleInterop": false,
-    "allowSyntheticDefaultImports": true,
+    "allowSyntheticDefaultImports": false,
```
Comment (arv) [inline ts-web-component/tsconfig.json:9]: Not sure if we have this as true in other places. It is a strange to allow this.

**Review by arv — COMMENTED** [review]
Hopefully you will find this useful.

File: ts-web-component/src/component/todo-item.ts (hunk 10 above)
Comment (arv) [reply ts-web-component/src/component/todo-item.ts:34]: try to store unnecessary state on the object

File: ts-web-component/src/component/my-todo.ts [hunk 19]
```diff
@@ -1,149 +1,155 @@
...
-      $item.todoID = item.id;
-      $item.addEventListener("onRemove", this.handleDeleteTodos.bind(this));
-      $item.addEventListener("onToggle", this.handleUpdateTodo.bind(this));
-      this.listContainer?.appendChild($item);
-    });
+  private _render() {
+    assert(this._listContainer);
+    assert(this._itemCount);
+    this._itemCount.textContent = `${this._list.length}`;
+    this._listContainer.textContent = "";
```
[the remark this answers is not in the document]
Comment (aboodman) [reply ts-web-component/src/component/my-todo.ts:128]: Another option is to use either watch with our own manual filter or some kind of state management. Maybe add a comment at least for the reader on some options?

**Comment by arv** [comment]
It would probably also be better to use shadow dom but I didn't want to go too far.

File: ts-web-component/src/component/my-todo.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (aboodman) [reply ts-web-component/src/component/my-todo.ts:128]: I think the code should definitely acknowledge that this isn't best practice and suggest some alternatives.

Here is one random idea though:

Could you stash the Replicache `entry` as a field on each todo item and compare by reference? Replace the item in its entirety if it has changed? Quick and easy.

[the remark this answers is not in the document]
Comment (aboodman) [reply ts-web-component/src/component/my-todo.ts:128]: I guess that is just what a vdom is for a list?

**Comment by cesara** [comment]
> This repo is missing eslint which should be setup and use the same config as replicache-internal. Maybe we should create a `@rocicorp/eslint` repo? A github repo should be sufficient. No need to publish to npm (I believe).
> 

I placed the esconfig inside the package.json I can split it out in its own file

**Comment by arv** [comment]
I don't mind having the eslint config in package.json. One less file seems good. But there is no devDependency on eslint, no lint script and the eslint config does not look correct/sufficient to me.

I would prefer if we could create a repo to share the config between all our projects

**Review by cesara — APPROVED** [review]
Nice, thanks for this, super helpful.

**Pull request merged**